From 88c22538853474a1ba6189105572a556c9bf7097 Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@foss.st.com>
Date: Mon, 15 Sep 2025 14:10:14 +0200
Subject: [PATCH] i3c: master: stm32: fix MTYPE for second part of I3C SDR
 direct CCC command

Direct CCC commands are split into two parts:
- the first part uses MTYPE=b0110 for the CCC command ID,
- the second part uses MTYPE=b0011 for the direct message with the target
  address.
Previously, MTYPE was not set for the second part, causing incorrect
message handling.

Fix the issue by specifying MTYPE=b0011 for the direct message in the
second part of the direct CCC command.

Change-Id: Ia6172d7809fe5bbf374db060e6a0a8af29f447b8
Fixes: 971dbeffd0e2 ("i3c: master: Add STM32 I3C Controller driver")
Signed-off-by: Amelie Delaunay <amelie.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/490664
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/i3c/master/stm32-i3c-master.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/i3c/master/stm32-i3c-master.c
+++ b/drivers/i3c/master/stm32-i3c-master.c
@@ -1477,7 +1477,8 @@ static int stm32_i3c_master_send_direct_
 
 	/* Directed message */
 	_cmd = &xfer->cmds[1];
-	_cmd->msg_cr |= FIELD_PREP(I3C_CR_ADD, cmd->dests[0].addr);
+	_cmd->msg_cr = FIELD_PREP(I3C_CR_MTYPE, MTYPE_I3C_DIRECT_MSG) |
+		       FIELD_PREP(I3C_CR_ADD, cmd->dests[0].addr);
 	_cmd->data_len = cmd->dests[0].payload.len;
 	_cmd->msg_cr |= FIELD_PREP(I3C_CR_DCNT, _cmd->data_len);
 	if (cmd->rnw) {
